File: google/compareStrings.py
import math

def getBiggerStrings(frequencies, key):

    # do binary search in frequencies for key
    first = 0
    last = len(frequencies)-1

    while first<=last:
        mid = math.floor((first+last)/2)

        if frequencies[mid]>key:
            last=mid-1
        elif frequencies[mid]<key:
            first=mid+1
        else:
            while mid>=0 and frequencies[mid]==key:
                mid = mid -1
            
            return mid+1
    
    if first>(len(frequencies)-1):
        return len(frequencies)
    elif last<0:
        return 0
    else:
        return last+1

if __name__=="__main__":
    first = "abcd,aabc,bd"
    second = "aaa,aa"

    # find sizes of all the strings in first
    firstSizes = [word.count(min(word)) for word in first.split(',')]
    
    # sort the first frequency array
    firstSizes.sort()

    #print(firstSizes)

    # for each word in second, compare its size to the sizes of strings in first 
    # by doing a binary search in the frequencies for that key
    result = []

    for word in second.split(','):
        size = word.count(min(word))

        result.append(getBiggerStrings(firstSizes,size))

    print(result)

File: google/compareStrings.cpp
#include <iostream>
#include <vector>
#include <sstream>


using namespace std;


class Solution
{
	public:

	int GetStringSize(const string & str1)
	{
		vector<char> frequencies(26,0);
		char smallest = str1[0];

		for (auto & curChr : str1)
		{
			if (curChr < smallest)
				smallest = curChr;

			frequencies[curChr-'a']++;
		}
		return static_cast<int>(frequencies[smallest-'a']);
	}
	
	vector<int> CompareStrings(const string& str1,const string& str2)
	{
		vector<string> vec1, vec2;
		string strTemp;
		istringstream iStream(str1);
		while(getline(iStream,strTemp,','))
			vec1.emplace_back(strTemp);
		
		iStream = istringstream(str2);
		while(getline(iStream,strTemp,','))
			vec2.emplace_back(strTemp);

		// compute the sizes of all strings in vec
		// keep frequency occurrences b/w 0 and 10 in array
		// calculate how many times each freq occurs finally make it cumulative
		
		vector<int> vecCumulative(11,0);

		for(auto & curString: vec1)
			vecCumulative[GetStringSize(curString)]++;

		for(int i=1; i<vecCumulative.size(); i++)
			vecCumulative[i] += vecCumulative[i-1];

		vector<int> vecRes;
		//calculate size of each string in vec2, compare to vecCumulative
		for (auto & curString: vec2)
		{
			int size = GetStringSize(curString);

			vecRes.emplace_back(size>0 ? vecCumulative[size-1] : 0);
		}
		return vecRes;

	}
};


int main()
{
	string str1 = "abcd,aabc,bd";
	string str2 = "aaa,aa";	

	Solution S1;
	for (auto & x: S1.CompareStrings(str1,str2))
		cout << x << " ";

	return 0;
}
